src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,211 @@
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(peerId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established as needed.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::LastUnicastBindingRemoved(FabricIndex fabric, NodeId node)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+    PeerId peer                      = fabricInfo->GetPeerIdForNode(node);
+    PendingNotificationEntry * entry = mPendingNotificationMap.FindEntry(peer);
+    if (entry)
+    {
+        mPendingNotificationMap.RemoveEntry(entry);
+    }
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(peer);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingManager::NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster, void * context)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        EmberBindingTableEntry entry;
+
+        if (emberGetBinding(i, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.local == endpoint &&
+            entry.clusterId == cluster)
+        {
+            if (entry.type == EMBER_UNICAST_BINDING)
+            {
+                FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(entry.fabricIndex);
+                VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+                PeerId peer                         = fabricInfo->GetPeerIdForNode(entry.nodeId);
+                OperationalDeviceProxy * peerDevice = mAppServer->GetCASESessionManager()->FindExistingSession(peer);
+                if (peerDevice != nullptr && mBoundDeviceChangedHandler)
+                {
+                    // We already have an active connection
+                    mBoundDeviceChangedHandler(&entry, peerDevice, NullOptional, context);
+                }
+                else
+                {
+                    // Enqueue pending cluster and establish connection
+                    ReturnErrorOnFailure(
+                        EnqueueUnicastNotification(entry.fabricIndex, entry.nodeId, entry.local, entry.clusterId, context));
+                    ReturnErrorOnFailure(EstablishConnection(entry.fabricIndex, entry.nodeId));
+                }
+            }
+            else if (entry.type == EMBER_MULTICAST_BINDING)
+            {
+                mBoundDeviceChangedHandler(&entry, nullptr, Optional<GroupId>(entry.groupId), context);
```

Comment:
Followup: if we're passing the whole entry, there's no reason to pass the entry's groupid...  Still need to sort out a sane signature for this callback.